### PR TITLE
Linear interpolation in ww3_gint adds energy

### PR DESCRIPTION
**Pull request by SanderHulst**
Hi,

Please find my update for ww3_gint. A description of the problem it tries to solve is found in #227 

Nearest interpolation is added by:
* reusing the EXTRA feature which accepts codes 0, 1 and 2. It is is now called INTMETHOD
* passing the INTMETHOD to the W3EXGI subroutine
* finding the point with the most weight
* set all weights except that one to -1 and the nearest to 1
* In the loop that multiplies valies by weight, early exit when weight not > 0.

I chose to modify the weights as late as possible so that the binary file still holds weights for linear interpolation and it is thus possible to switch between interpolation methods without the need to remove the file.

### CODE REVIEW

**Comment by SanderHulst** [comment]
Something odd happened with the rebase. Somehow redid commits too. If this PR is approved (..) "Squash and merge" will fix this.

**Comment by SanderHulst** [comment]
The changes do not perform as expected. PR set to work in progress.

**Comment by aliabdolali** [comment]
Reference issue #[227](https://github.com/NOAA-EMC/WW3/issues/227)

Regression test results are attached.
[regtest summary on NCEP HPC](https://github.com/NOAA-EMC/WW3/files/6117664/Regtest_Comp_NCEP_Nearest.zip)

No update to the existing regtests. The tests with ww3_gint have expected differences. 

The other differences are all in regtests that have existing known differences:

mww3_test_03
ww3_tp2.18

**Review by JessicaMeixner-NOAA — CHANGES_REQUESTED** [review]
These are great updates. Thank you!  The input file description in the model/inp/ww3_gint.inp needs to be updated to reflect the changes that are here. Are there also manual updates that should also be made?  Lastly, I'm confused by the answer changes in the gint test as this just seems like an extra option not changing answers to current options.  That being said, we should update or add to one of the regression tests to test this option so it's kept functioning.

**Comment by aliabdolali** [comment]
Hi @SanderHulst 
Could you update model/inp/ww3.gint.inp? and manual if applicable?

**Comment by SanderHulst** [comment]
Yes, I'll try to add something by the end of the week. No other actions requested from my side, right?

**Comment by aliabdolali** [comment]
> These are great updates. Thank you! The input file description in the model/inp/ww3_gint.inp needs to be updated to reflect the changes that are here. Are there also manual updates that should also be made? Lastly, I'm confused by the answer changes in the gint test as this just seems like an extra option not changing answers to current options. That being said, we should update or add to one of the regression tests to test this option so it's kept functioning.

@JessicaMeixner-NOAA the standard output of ww3_gint.out is changed, due to changes in the definition, so they are expected. Sander is taking care of manual and model/inp/ww3_gint.inp updates.

**Comment by aliabdolali** [comment]
> Yes, I'll try to add something by the end of the week. No other actions requested from my side, right?

@SanderHulst Awesome, we can proceed once you updated the model/inp/ww3_gint.inp and the manual (please search for gint and update those sections).

**Comment by SanderHulst** [comment]
Something like this?

File: model/inp/ww3_gint.inp [hunk 1]
```diff
@@ -28,7 +28,10 @@ $ Upon execution of the code an out_grd.grd4 is generated via
 $ interpolation of output fields from the various out_grd.grdN
 $ (N varying from 1 to 3) files.
 $
-$ Choice of no extrapolation [0] or to extrapolate [1] with wet points [2]
+$ Interpolation scheme
+$   0: Linear
```
Comment (JessicaMeixner-NOAA) [inline model/inp/ww3_gint.inp]: Thanks for adding the update here and in the manual! This is exactly what I was mentioning. 

I have one question though, based on what I've looked at the code and what was there before and that we've only added 1 new option (option 2) I thought the options would be more described like: 

```
0: linear interpolation, no extrapolation 
1: linear interpolation, extrapolate with wet points 
2: nearest neighbor interpolation 
```

Or am I misunderstanding?

Comment (SanderHulst) [reply model/inp/ww3_gint.inp]: From what I gathered, 1 only works for unstructured grids. I did not check whether that uses some sort of linear interpolation (baricentric coordinates?)

Comment (SanderHulst) [reply model/inp/ww3_gint.inp]: Yes, it does.

Comment (JessicaMeixner-NOAA) [reply model/inp/ww3_gint.inp]: I don't see where 1 only works for unstructured grids?   I can't find anywhere where it's limited based on grid type.  @aliabdolali you work with unstructured grids more than I do, do you know about this?   

@aliabdolali also, all regression tests only use the 0 option : none use 1 or 2 (which is the new feature added here), from what I can find.

Comment (aliabdolali) [reply model/inp/ww3_gint.inp]: @JessicaMeixner-NOAA I never used ww3_gint in my unstructured applications. 
Do you suggest having a new regtest with options 1 and 2 ? if yes,  I can add them.

**Comment by JessicaMeixner-NOAA** [comment]
I've looked at this again today, and again I do not see restriction on the grid type "GTYPE" for which method is chosen in ww3_gint.  @SanderHulst if you could provide additional reasoning of how you see the restriction of  1 only working for unstructured grids that would be very helpful.  Otherwise, perhaps another way to determine this is that we do have ww3_gint in some of the regression tests with structured grids.  We could set that to 1 and see if it works or not.   

I don't have any issues with the code additions itself.  The only thing I am still unsure about is the description for option 1 and whether there is actually a limitation to unstructured grids.   I do think we should add an issue to add regression tests to test this new feature if it's not added before merging.   During the review of this PR we discovered that no regression test currently is using the linear interpolation.  I've added an issue to track this as a separate issue: https://github.com/NOAA-EMC/WW3/issues/351 as it does not need to be addressed by this PR.

**Comment by aliabdolali** [comment]
Hi @SanderHulst 
Could you review the PR I made to your feature branch? 
I also asked a couple of questions. I'd appreciate it if you answer them, so we can proceed and reintegrate your developments to the main repo. Thanks for your cooperation.

**Comment by aliabdolali** [comment]
Hi @SanderHulst I was trying to check the ww3_gint options (0,1,2) on structured and unstructured grids. 
What I did so far is the following:
For the structured grids (mww3_test_03, grd_set_a). I used ww3_gint with 0, 1 and 2 options and plotted them. They are identical. 
For the unstructured grids (ww3_tp2.21, grd_set_a). I used ww3_gint with 0, 1 and 2 options and plotted them. They are identical. 
i wonder if I am doing sth wrong or I should expect zero difference for these two cases?

**Review by aliabdolali — APPROVED** [review]
I am going to accept this PR, and will open an issue to test new regtests to check these new features (gint interpolation options).

**Review by aliabdolali — APPROVED** [review]

**Review by aliabdolali — APPROVED** [review]

**Review by JessicaMeixner-NOAA — APPROVED** [review]
Any follow up can occur in subsequent PRs.
